src/hostnet/forwards.ml

```diff
@@ -0,0 +1,487 @@
+  | Ezjsonm.Parse_error (_v, msg) -> Error (`Msg msg)
+  | e -> Error (`Msg (Printf.sprintf "parsing %s: %s" x (Printexc.to_string e)))
+
+let dynamic = ref []
+let static = ref []
+let all = ref []
+
+let set_static xs =
+  static := xs;
+  all := !static @ !dynamic;
+  Log.info (fun f -> f "New Forward configuration: %s" (to_string !all))
+
+let update xs =
+  dynamic := xs;
+  all := !static @ !dynamic;
+  Log.info (fun f -> f "New Forward configuration: %s" (to_string !all))
+
+(* Extend a SHUTDOWNABLE flow with a `read_some` API, as used by "channel".
+   Ideally we would use channel, but we need to access the underlying flow
+   without leaving data trapped in the buffer. *)
+module type Read_some = sig
+  include Mirage_flow_combinators.SHUTDOWNABLE
+
+  val read_some :
+    flow -> int -> (Cstructs.t Mirage_flow.or_eof, error) result Lwt.t
+end
+
+module Read_some (FLOW : Mirage_flow_combinators.SHUTDOWNABLE) : sig
+  include Read_some
+
+  val connect : FLOW.flow -> flow
+end = struct
+  (* A flow with a buffer, filled by "read_some" and then drained by "read" *)
+  type flow = { mutable remaining : Cstruct.t; flow : FLOW.flow }
+
+  let connect flow = { remaining = Cstruct.create 0; flow }
+
+  type error = FLOW.error
+
+  let pp_error = FLOW.pp_error
+
+  type write_error = FLOW.write_error
+
+  let pp_write_error = FLOW.pp_write_error
+
+  let read_some flow len =
+    let open Lwt.Infix in
+    let rec loop acc len =
+      if Cstruct.length flow.remaining = 0 && len > 0 then
+        FLOW.read flow.flow >>= function
+        | Error e -> Lwt.return (Error e)
+        | Ok (`Data buf) ->
+            flow.remaining <- buf;
+            loop acc len
+        | Ok `Eof -> Lwt.return (Ok `Eof)
+      else if Cstruct.length flow.remaining < len then (
+        let take = flow.remaining in
+        flow.remaining <- Cstruct.create 0;
+        loop (take :: acc) (len - Cstruct.length take))
+      else
+        let take, leave = Cstruct.split flow.remaining len in
+        flow.remaining <- leave;
+        Lwt.return @@ Ok (`Data (List.rev (take :: acc)))
+    in
+    loop [] len
+
+  let read flow =
+    if Cstruct.length flow.remaining > 0 then (
+      let result = flow.remaining in
+      flow.remaining <- Cstruct.create 0;
+      Lwt.return @@ Ok (`Data result))
+    else FLOW.read flow.flow
+
+  let write flow = FLOW.write flow.flow
+  let writev flow = FLOW.writev flow.flow
+  let close flow = FLOW.close flow.flow
+  let shutdown_write flow = FLOW.shutdown_write flow.flow
+  let shutdown_read flow = FLOW.shutdown_read flow.flow
+end
+
+module Handshake (FLOW : Read_some) = struct
+  (* A Message is a buffer prefixed with a uint16 length field. *)
+  module Message = struct
+    open Lwt.Infix
+
+    let pp_error ppf = function
+      | `Flow e -> FLOW.pp_error ppf e
+      | `Eof -> Fmt.string ppf "EOF while reading handshake"
+
+    let read flow =
+      FLOW.read_some flow 2 >>= function
```

Comment:
minor, but I think you could replace this `>>=` with `>!=` and lose all the `Lwt.return`s below.